File: Python/MergeSortedLists.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def merge_two_lists(list1: ListNode, list2: ListNode) -> ListNode:
    dummy = ListNode(0)
    tail = dummy

    while list1 and list2:
        if list1.val <= list2.val:
            tail.next = list1
            list1 = list1.next
        else:
            tail.next = list2
            list2 = list2.next
        tail = tail.next

    if list1:
        tail.next = list1
    if list2:
        tail.next = list2

    return dummy.next

# Helper to print list
def print_list(head: ListNode):
    while head:
        print(head.val, end=" -> ")
        head = head.next
    print("None")

# Example 1
list1 = ListNode(2, ListNode(4, ListNode(7, ListNode(9))))
list2 = ListNode(1, ListNode(2, ListNode(5, ListNode(6))))
merged1 = merge_two_lists(list1, list2)
print_list(merged1)  # Output: 1 -> 2 -> 2 -> 4 -> 5 -> 6 -> 7 -> 9 -> None

# Example 2
l1 = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
l2 = ListNode(5, ListNode(6, ListNode(10)))
merged2 = merge_two_lists(l1, l2)
print_list(merged2)  # Output: 1 -> 2 -> 3 -> 4 -> 5 -> 6 -> 10 -> None

File: JAVA/MergeSortedLists.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int val) { this.val = val; this.next = null; }
}

public class MergeSortedLists {

    public static ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode dummy = new ListNode(0);
        ListNode tail = dummy;

        while (list1 != null && list2 != null) {
            if (list1.val <= list2.val) {
                tail.next = list1;
                list1 = list1.next;
            } else {
                tail.next = list2;
                list2 = list2.next;
            }
            tail = tail.next;
        }

        if (list1 != null) tail.next = list1;
        if (list2 != null) tail.next = list2;

        return dummy.next;
    }

    // Helper to print list
    public static void printList(ListNode head) {
        while (head != null) {
            System.out.print(head.val + " -> ");
            head = head.next;
        }
        System.out.println("null");
    }

    public static void main(String[] args) {
        // Example 1
        ListNode list1 = new ListNode(2);
        list1.next = new ListNode(4);
        list1.next.next = new ListNode(7);
        list1.next.next.next = new ListNode(9);

        ListNode list2 = new ListNode(1);
        list2.next = new ListNode(2);
        list2.next.next = new ListNode(5);
        list2.next.next.next = new ListNode(6);

        ListNode merged1 = mergeTwoLists(list1, list2);
        printList(merged1); // Output: 1 -> 2 -> 2 -> 4 -> 5 -> 6 -> 7 -> 9 -> null

        // Example 2
        ListNode l1 = new ListNode(1);
        l1.next = new ListNode(2);
        l1.next.next = new ListNode(3);
        l1.next.next.next = new ListNode(4);

        ListNode l2 = new ListNode(5);
        l2.next = new ListNode(6);
        l2.next.next = new ListNode(10);

        ListNode merged2 = mergeTwoLists(l1, l2);
        printList(merged2); // Output: 1 -> 2 -> 3 -> 4 -> 5 -> 6 -> 10 -> null
    }
}
